nemo/collections/nlp/models/language_modeling/megatron/gpt_full_te_layer_autocast_spec.py

```diff
@@ -0,0 +1,273 @@
+        )
+        self.is_first_microbatch = False
+        context = None
+
+        return hidden_states, context
+
+    def _get_layer_offset(self):
+
+        pipeline_rank = parallel_state.get_pipeline_model_parallel_rank()
+
+        num_layers_per_pipeline_rank = (
+            self.config.num_layers // parallel_state.get_pipeline_model_parallel_world_size()
+        )
+
+        if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+            vp_rank = parallel_state.get_virtual_pipeline_model_parallel_rank()
+            vp_size = parallel_state.get_virtual_pipeline_model_parallel_world_size()
+
+            total_num_layers = self.config.num_layers
+            num_layers_per_virtual_rank = num_layers_per_pipeline_rank // vp_size
+            total_virtual_chunks = total_num_layers // vp_size
+            offset = vp_rank * total_virtual_chunks + (pipeline_rank * num_layers_per_virtual_rank)
+
+        else:
+            # Each stage gets a contiguous set of layers.
+            if parallel_state.get_pipeline_model_parallel_world_size() > 1:
+                offset = pipeline_rank * num_layers_per_pipeline_rank
+            else:
+                offset = 0
+
+        return offset
+
+    def sharded_state_dict(self, prefix: str = '', sharded_offsets: tuple = ()):
+        TENSOR_PARALLEL_LAYERS_AXIS_MAP = {
+            'self_attention.layernorm_qkv.weight': 0,
+            'self_attention.layernorm_qkv.bias': 0,
+            "self_attention.proj.weight": 1,
+            "layernorm_mlp.fc1_weight": 0,
+            "layernorm_mlp.fc1_bias": 0,
+            "layernorm_mlp.fc2_weight": 1,
+        }
+
+        state_dict = self.state_dict(prefix='', keep_vars=True)
+        sharded_state_dict = make_sharded_tensors_for_checkpoint(
+            state_dict, prefix, TENSOR_PARALLEL_LAYERS_AXIS_MAP, sharded_offsets
+        )
+
+        # TODO: we need to add sharded_state_dict_keys_map to the config. Like in TransformerLayer submodules config
+        # prefixed_map = {
+        #    f'{prefix}{k}': f'{prefix}{v}'
+        #    for k, v in self.config.sharded_state_dict_keys_map.items()
+        # }
+
+        # if prefixed_map:
+        #    apply_prefix_mapping(sharded_state_dict, prefixed_map)
+
+        return sharded_state_dict
+
+
+# Use this spec to use the full Transformer layer from Transformer Engine
+def get_gpt_full_te_layer_autocast_spec() -> TransformerBlockSubmodules:
+    if not HAVE_MEGATRON_CORE:
+        raise ImportError(
+            "megatron-core was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+        )
+    if not HAVE_TE:
+        raise ImportError(
+            "TransformerEngine was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+        )
+
+    return TransformerBlockSubmodules(layer_specs=ModuleSpec(module=TETransformerLayerAutocast))
```

Comment:
Could you try this instead? You'll need to use a different branch from mcore to test this 
```suggestion
    return ModuleSpec(module=TETransformerLayerAutocast)
```

Reply:
Same here, we need this change as the MCore [PR](https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/1113) has been merged.